### PR TITLE
[homework04] Implement an authentication & authorization with the JWT.

### PR DESCRIPTION
**Pull request by ArzamastsevOleksandr**

### CODE REVIEW

File: homework02/src/routes/user-routes.js [hunk 1]
```diff
@@ -1,27 +1,29 @@
...
-    winston.info(`Method: ${req.method}. URL: ${req.url}. Params: ${JSON.stringify(req.params)}`);
-    next();
-};
+const validateRequest = Validator(false);
 
 module.exports = (app) => {
     const users = require('../controllers/user-controller');
 
-    app.get('/', log, users.greeting);
+    app.post('/login', tokenGenerator.login);
```
Comment (DiannaDea) [inline homework02/src/routes/user-routes.js]: TODO: split auth and users routes. Now you have all routes defined in `user-routes.js`

So all auth routes have to start with `/auth`: `/auth/login`, `/auth/refresh`
Users routes start with `/users`

File: homework02/src/authentication/tokenGenerator.js [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+        } else {
+            res.status(400).end(util.toJson(util.errorData('Authentication failed! Please check the request')));
+        }
+    }
+
+    refresh(req, res) {
+        const body = req.body;
+
+        if (body.refreshToken) {
+            const userId = db.findUserIdWithLoginAndPassword(body.username, body.password);
```
Comment (DiannaDea) [inline homework02/src/authentication/tokenGenerator.js]: TODO: Don't pass `username` and `password` when refreshing tokens. We pass `accessToken` and `refreshToken` in `req.body`

File: homework02/src/authentication/tokenGenerator.js [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+            if (userId) {
+                const oldTokens = db.getTokensForId(userId);
+                const oldRefreshToken = oldTokens.refreshToken;
+
+                if (body.refreshToken === oldRefreshToken) {
+                    const newToken = jwt.sign({username: body.username, password: body.password}, config.secret, {expiresIn: config.tokenLife})
+                    db.updateTokensForId(userId, newToken, body.refreshToken);
```
Comment (DiannaDea) [inline homework02/src/authentication/tokenGenerator.js]: TODO: Create new tokens pair: `refresh and access`, not only `access`

**Comment by DiannaDea** [comment]
Please have a look at my implementation of auth flow. Ask questions if needed. 

```

const checkTokenValidity = (token) => {
  try {
    jwt.verify(token, PUBLIC_KEY, {algorithms: [hash_algorithm]})
    return true
  } catch (error) {
    logger.error('Token expired')
    return false
  }
}

const AuthController = {
  login: async (ctx) => {
    const {username, password} = ctx.request.body

    const user = await User.findOne({username})

    if (!user || !await comparePasswords(password, user.password)) {
      return ctx.send(400, {
        errors: ['Incorrect username or password']
      })
    }

    const accessTokenIsValid = checkTokenValidity(user.accessToken)
    const refreshTokenIsValid = checkTokenValidity(user.refreshToken)

    if (accessTokenIsValid && refreshTokenIsValid) {
      return ctx.send(200, {
        access_token: user.accessToken,
        refresh_token: user.refreshToken
      })
    }

    const newTokenPair = generateTokensPair(
        user._id,
        user.username,
        user.roles
    )

    await User.findByIdAndUpdate(user._id, {...newTokenPair})

    return ctx.send(200, {
      access_token: newTokenPair.accessToken,
      refresh_token: newTokenPair.refreshToken
    })
  },
  refreshTokens: async (ctx) => {
    const {access_token, refresh_token} = ctx.request.body

    const {sub} = decodeToken(refresh_token)
    const user = await User.findById(sub)

    if (!user) {
      return ctx.send(404, {
        errors: ['No user found']
      })
    }

    if (user.refreshToken !== refresh_token ||
      user.accessToken !== access_token
    ) {
      return ctx.send(400, {
        errors: ['Incorrect access or refresh token']
      })
    }

    const newTokenPair = generateTokensPair(sub, user.username, user.roles)
    await User.findByIdAndUpdate(sub, {...newTokenPair})

    return ctx.send(200, {
      access_token: newTokenPair.accessToken,
      refresh_token: newTokenPair.refreshToken
    })
  }
}
```

File: homework02/logs/info.log [hunk 4]
```diff
@@ -78,3 +78,185 @@
 {"message":"Method: POST. URL: /users. Params: {}","level":"info"}
 {"message":"Server is up and running on port 3000","level":"info"}
 {"message":"Method: POST. URL: /users. Params: {}","level":"info"}
+{"message":"Server is up and running on port 3000","level":"info"}
```
[the remark this answers is not in the document]
Comment (ArzamastsevOleksandr) [reply homework02/logs/info.log]: Added this to .gitignore.

File: homework02/src/authentication/auth.js [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+
+const jwt = require('jsonwebtoken');
+const config = require('../../config/jwt');
+const Util = require('../util/util');
+
+const util = new Util();
+
+const extractToken = req => req.headers['x-access-token'] || req.headers['authorization'];
+
+const removeRedundantParts = token => token.startsWith('Bearer ') ? token.slice(7, token.length) : token;
```
[the remark this answers is not in the document]
Comment (ArzamastsevOleksandr) [reply homework02/src/authentication/auth.js]: I understand, that, for example, it is true with complex expressions, where a lot of possible combinations is involved in only one expression. Writing your own parser would only waste time. But in this case I chose simplicity over reusability. To create a simple pattern and extract the Bearer from the HTTP request one would have to write the following (or smth like that): 

**const bearerPattern = /[A-Za-z0-9\-\._~\+\/]+=*/;
const removeRedundantParts = token => token.replace(bearerPattern, '');**

It is definitely more difficult to understand and manage, so I decided to stick to a simple solution.

File: homework02/src/authentication/tokenGenerator.js [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+const DB = require('../in-memory-db/in-memory-db');
+const Util = require('../util/util');
+
+const db = new DB().getInstance();
+const util = new Util();
+
+class TokenGenerator {
+    login(req, res) {
+        const body = req.body;
+        const username = body.username;
```
[the remark this answers is not in the document]
Comment (ArzamastsevOleksandr) [reply homework02/src/authentication/tokenGenerator.js]: Done.

File: homework02/src/authentication/tokenGenerator.js [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+        const password = body.password;
+
+        if (username && password) {
+            const userId = db.findUserIdWithLoginAndPassword(username, password);
+
+            if (userId) {
+                const newToken = jwt.sign({username: username, password: password}, config.secret, {expiresIn: config.tokenLife});
```
[the remark this answers is not in the document]
Comment (ArzamastsevOleksandr) [reply homework02/src/authentication/tokenGenerator.js]: Understood. Done.

File: homework02/src/authentication/tokenGenerator.js [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+
+                res.status(200).end(util.toJson({
+                    message: 'Authentication successful!',
+                    token: newToken,
+                    refreshToken: refreshToken
+                }));
+            } else {
+                res.status(403).end(util.toJson(util.errorData('Incorrect login or password')));
+            }
+        } else {
```
[the remark this answers is not in the document]
Comment (ArzamastsevOleksandr) [reply homework02/src/authentication/tokenGenerator.js]: A very good point! Solved this with method decomposition to make each method tackle one step at a time.

File: homework02/src/in-memory-db/in-memory-db.js [hunk 9]
```diff
@@ -52,6 +53,23 @@ class InMemoryDb {
             .sort(comparator)
             .slice(0, limit);
     }
+
+    findUserIdWithLoginAndPassword(login, password) {
+        const result = this.fetchAllUsers().filter(user => user.login === login && user.password === password);
+
+        return result.length > 0 ? result[0].id : null;
```
[the remark this answers is not in the document]
Comment (ArzamastsevOleksandr) [reply homework02/src/in-memory-db/in-memory-db.js]: Done.

File: homework02/src/routes/user-routes.js (hunk 1 above)
Comment (ArzamastsevOleksandr) [reply homework02/src/routes/user-routes.js]: Done.

File: homework02/src/authentication/tokenGenerator.js (hunk 3 above)
Comment (ArzamastsevOleksandr) [reply homework02/src/authentication/tokenGenerator.js]: Done.

File: homework02/src/authentication/tokenGenerator.js (hunk 2 above)
Comment (ArzamastsevOleksandr) [reply homework02/src/authentication/tokenGenerator.js]: Fixed.

**Pull request merged**